### DISCUSSION TITLE
What is the best idiomatic way to update a dashboard made in fltk rust ?

### DISCUSSION THREAD

**Discussion by vortex314** [fltk-rs/fltk-rs#1426] [Q&A]
I'm trying to build a dashboard that subscribes to frequently published data in a PubSub system like MQTT or Redis.

I'm using now the app.wait cycle to receive the events and update the GUI. I notice however this is unpredictable in reaction time if the cursor is not in the GUI.

Is there a better way ? To make this more real time ? 

```rust
        win.end();
        win.show();
        // sleep
        let sub = rx_publish.resubscribe();
        while _app.wait() {
            info!("app wait");
            if let Ok(x) = rx_publish.try_recv() {
                match x {
                    RedisEvent::Publish(PublishMessage { topic, message }) => {
                        _but.set_label(topic.as_str())
                    }
                    _ => {}
                }
            }
        }
```

**Comment by vortex314** [comment]
Ok, I could make it respond as expected by just calling awake()  at the moment an event is send.
Better ways to do so ? 
```rust
while let Some(msg) = pubsub_stream.next().await {
            info!(
                "|{:>20}| Redis topic: {}",
                thread::current().name().unwrap(),
                msg.get_channel_name().to_string(),
            );
            awake();
            match tx_broadcast.send(RedisEvent::Publish(PublishMessage {
                topic: msg.get_channel_name().to_string(),
                message: msg.get_payload().unwrap(),
            })) {
                Ok(_) => {}
                Err(e) => {
                    error!("|{:>20}| error: {}", thread::current().name().unwrap(), e);
                    break;
                }
            }
        }
```

**Reply by MoAlyousef** [reply]
I think awake in the event loop is the way to go. Any attempt to modify the gui from another execution context (other thread or task in this case) will require a call to app::awake().
The other way is to spawn a detached thread where you listen for changes, but it will also require an awake call. 
The probably worst option is to use a timer and recheck at every interval for change.

**Reply by vortex314** [reply]
Thanks for the answer , confirms my direction of coding.
Thanks also a lot for your immense and relentless effort to develop this and answer a lot of support questions !
I like your Youtube explanatory videos also.

**Reply by MoAlyousef** [reply]
Thank you for your comment. I really appreciate it :)
